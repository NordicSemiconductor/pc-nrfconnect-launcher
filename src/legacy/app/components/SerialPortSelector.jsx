/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import React from 'react';
import Dropdown from 'react-bootstrap/Dropdown';
import { Iterable } from 'immutable';
import { bool, func, instanceOf, oneOfType, string } from 'prop-types';

import HotkeyedDropdown from '../../components/HotkeyedDropdown';
import portPath from '../../portPath';

const SEGGER_VENDOR_IDS = new Set(['0x1366', '1366']);

function filterSeggerPorts(port) {
    return SEGGER_VENDOR_IDS.has(port.vendorId);
}

class SerialPortSelector extends React.Component {
    renderSerialPortItems() {
        const { ports, onSelect, isLoading, menuItemCssClass, filter } =
            this.props;

        if (!isLoading) {
            return ports.filter(filter).map(port => (
                <Dropdown.Item
                    key={portPath(port)}
                    className={menuItemCssClass}
                    eventKey={portPath(port)}
                    onSelect={() => onSelect(port)}
                >
                    <div>{portPath(port)}</div>
                    <div style={{ fontSize: 'small' }}>
                        {port.serialNumber || ''}
                    </div>
                </Dropdown.Item>
            ));
        }
        return null;
    }

    renderCloseItem() {
        const { selectedPort, isLoading, onDeselect, menuItemCssClass } =
            this.props;

        if (selectedPort && !isLoading) {
            return (
                <Dropdown.Item
                    className={menuItemCssClass}
                    eventKey="Close serial port"
                    onSelect={onDeselect}
                >
                    <div>Close serial port</div>
                </Dropdown.Item>
            );
        }
        return null;
    }

    render() {
        const {
            selectedPort,
            showPortIndicator,
            portIndicatorStatus,
            onToggle,
            isExpanded,
            cssClass,
        } = this.props;

        const selectorText = selectedPort || 'Select serial port';
        const indicatorCssClass = `core-serial-port-indicator ${portIndicatorStatus}`;

        return (
            <div className={cssClass}>
                <HotkeyedDropdown
                    id="serial-port-selector"
                    open={isExpanded}
                    onToggle={onToggle}
                    hotkey="Alt+P"
                    title={`${selectorText}`}
                >
                    {this.renderSerialPortItems()}
                    {this.renderCloseItem()}
                </HotkeyedDropdown>
                {showPortIndicator ? (
                    <div className={indicatorCssClass} />
                ) : (
                    <div />
                )}
            </div>
        );
    }
}

SerialPortSelector.propTypes = {
    ports: oneOfType([instanceOf(Array), instanceOf(Iterable)]).isRequired,
    selectedPort: string,
    showPortIndicator: bool,
    portIndicatorStatus: string,
    isLoading: bool,
    isExpanded: bool,
    onToggle: func.isRequired,
    onSelect: func.isRequired,
    onDeselect: func.isRequired,
    cssClass: string,
    menuItemCssClass: string,
    filter: func,
};

SerialPortSelector.defaultProps = {
    selectedPort: '',
    showPortIndicator: true,
    isExpanded: false,
    isLoading: false,
    portIndicatorStatus: 'off',
    cssClass: 'core-padded-row core-serial-port-selector',
    menuItemCssClass: 'btn-primary',
    filter: filterSeggerPorts,
};

export default SerialPortSelector;
