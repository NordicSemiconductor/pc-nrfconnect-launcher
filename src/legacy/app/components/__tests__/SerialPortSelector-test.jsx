/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

/* eslint-disable import/first */

// Do not render react-bootstrap components in tests
jest.mock('react-bootstrap', () => ({
    Dropdown: 'Dropdown',
    MenuItem: 'MenuItem',
    DropdownToggle: 'DropdownToggle',
    DropdownMenu: 'DropdownMenu',
}));

// Do not decorate components
jest.mock('../../../decoration', () => ({
    decorate: component => component,
}));

import React from 'react';
import renderer from 'react-test-renderer';

import SerialPortSelector from '../SerialPortSelector';

const SEGGER_VENDOR_ID = '0x1366';

describe('SerialPortSelector', () => {
    it('should render empty port list', () => {
        expect(
            renderer.create(
                <SerialPortSelector
                    ports={[]}
                    onToggle={() => {}}
                    onSelect={() => {}}
                    onDeselect={() => {}}
                    bindHotkey={() => {}}
                />
            )
        ).toMatchSnapshot();
    });

    it('should render two ports', () => {
        expect(
            renderer.create(
                <SerialPortSelector
                    ports={[
                        {
                            path: '/dev/tty0',
                            serialNumber: '123456',
                            vendorId: SEGGER_VENDOR_ID,
                        },
                        {
                            path: '/dev/tty1',
                            serialNumber: '456789',
                            vendorId: SEGGER_VENDOR_ID,
                        },
                    ]}
                    onToggle={() => {}}
                    onSelect={() => {}}
                    onDeselect={() => {}}
                    bindHotkey={() => {}}
                />
            )
        ).toMatchSnapshot();
    });

    it('should render only segger ports by default filter', () => {
        expect(
            renderer.create(
                <SerialPortSelector
                    ports={[
                        {
                            path: '/dev/tty0',
                            serialNumber: '123456',
                            vendorId: '0x1234',
                        },
                        {
                            path: '/dev/tty1',
                            serialNumber: '456789',
                            vendorId: SEGGER_VENDOR_ID,
                        },
                    ]}
                    onToggle={() => {}}
                    onSelect={() => {}}
                    onDeselect={() => {}}
                    bindHotkey={() => {}}
                />
            )
        ).toMatchSnapshot();
    });

    it('should render filtered ports', () => {
        expect(
            renderer.create(
                <SerialPortSelector
                    ports={[
                        {
                            path: '/dev/tty0',
                            serialNumber: '123456',
                            vendorId: '0x1234',
                        },
                        {
                            path: '/dev/tty1',
                            serialNumber: '456789',
                            vendorId: SEGGER_VENDOR_ID,
                        },
                    ]}
                    onToggle={() => {}}
                    onSelect={() => {}}
                    onDeselect={() => {}}
                    bindHotkey={() => {}}
                    filter={port => port.serialNumber === '123456'}
                />
            )
        ).toMatchSnapshot();
    });

    it('should render two ports, one selected', () => {
        expect(
            renderer.create(
                <SerialPortSelector
                    ports={[
                        {
                            path: '/dev/tty0',
                            serialNumber: '123456',
                            vendorId: SEGGER_VENDOR_ID,
                        },
                        {
                            path: '/dev/tty1',
                            serialNumber: '456789',
                            vendorId: SEGGER_VENDOR_ID,
                        },
                    ]}
                    selectedPort="/dev/tty1"
                    onToggle={() => {}}
                    onSelect={() => {}}
                    onDeselect={() => {}}
                    bindHotkey={() => {}}
                />
            )
        ).toMatchSnapshot();
    });

    it('should render empty list while loading', () => {
        expect(
            renderer.create(
                <SerialPortSelector
                    ports={[
                        {
                            path: '/dev/tty0',
                            serialNumber: '123456',
                            vendorId: SEGGER_VENDOR_ID,
                        },
                    ]}
                    isLoading
                    onToggle={() => {}}
                    onSelect={() => {}}
                    onDeselect={() => {}}
                    bindHotkey={() => {}}
                />
            )
        ).toMatchSnapshot();
    });

    it('should render empty list while loading and port is selected', () => {
        expect(
            renderer.create(
                <SerialPortSelector
                    ports={[
                        {
                            path: '/dev/tty0',
                            serialNumber: '123456',
                            vendorId: SEGGER_VENDOR_ID,
                        },
                    ]}
                    selectedPort="/dev/tty0"
                    isLoading
                    onToggle={() => {}}
                    onSelect={() => {}}
                    onDeselect={() => {}}
                    bindHotkey={() => {}}
                />
            )
        ).toMatchSnapshot();
    });
});
