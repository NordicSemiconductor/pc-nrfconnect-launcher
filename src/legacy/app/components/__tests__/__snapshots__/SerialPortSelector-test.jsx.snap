// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SerialPortSelector should render empty list while loading 1`] = `
<div
  className="core-padded-row core-serial-port-selector"
>
  <div
    className="dropdown"
    disabled={false}
    id="serial-port-selector"
    open={false}
  >
    <button
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-toggle btn btn-primary"
      disabled={false}
      onClick={[Function]}
      title="Select serial port (Alt+P)"
      type="button"
    >
      Select serial port
    </button>
  </div>
  <div
    className="core-serial-port-indicator off"
  />
</div>
`;

exports[`SerialPortSelector should render empty list while loading and port is selected 1`] = `
<div
  className="core-padded-row core-serial-port-selector"
>
  <div
    className="dropdown"
    disabled={false}
    id="serial-port-selector"
    open={false}
  >
    <button
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-toggle btn btn-primary"
      disabled={false}
      onClick={[Function]}
      title="/dev/tty0 (Alt+P)"
      type="button"
    >
      /dev/tty0
    </button>
  </div>
  <div
    className="core-serial-port-indicator off"
  />
</div>
`;

exports[`SerialPortSelector should render empty port list 1`] = `
<div
  className="core-padded-row core-serial-port-selector"
>
  <div
    className="dropdown"
    disabled={false}
    id="serial-port-selector"
    open={false}
  >
    <button
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-toggle btn btn-primary"
      disabled={false}
      onClick={[Function]}
      title="Select serial port (Alt+P)"
      type="button"
    >
      Select serial port
    </button>
  </div>
  <div
    className="core-serial-port-indicator off"
  />
</div>
`;

exports[`SerialPortSelector should render filtered ports 1`] = `
<div
  className="core-padded-row core-serial-port-selector"
>
  <div
    className="dropdown"
    disabled={false}
    id="serial-port-selector"
    open={false}
  >
    <button
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-toggle btn btn-primary"
      disabled={false}
      onClick={[Function]}
      title="Select serial port (Alt+P)"
      type="button"
    >
      Select serial port
    </button>
  </div>
  <div
    className="core-serial-port-indicator off"
  />
</div>
`;

exports[`SerialPortSelector should render only segger ports by default filter 1`] = `
<div
  className="core-padded-row core-serial-port-selector"
>
  <div
    className="dropdown"
    disabled={false}
    id="serial-port-selector"
    open={false}
  >
    <button
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-toggle btn btn-primary"
      disabled={false}
      onClick={[Function]}
      title="Select serial port (Alt+P)"
      type="button"
    >
      Select serial port
    </button>
  </div>
  <div
    className="core-serial-port-indicator off"
  />
</div>
`;

exports[`SerialPortSelector should render two ports 1`] = `
<div
  className="core-padded-row core-serial-port-selector"
>
  <div
    className="dropdown"
    disabled={false}
    id="serial-port-selector"
    open={false}
  >
    <button
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-toggle btn btn-primary"
      disabled={false}
      onClick={[Function]}
      title="Select serial port (Alt+P)"
      type="button"
    >
      Select serial port
    </button>
  </div>
  <div
    className="core-serial-port-indicator off"
  />
</div>
`;

exports[`SerialPortSelector should render two ports, one selected 1`] = `
<div
  className="core-padded-row core-serial-port-selector"
>
  <div
    className="dropdown"
    disabled={false}
    id="serial-port-selector"
    open={false}
  >
    <button
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-toggle btn btn-primary"
      disabled={false}
      onClick={[Function]}
      title="/dev/tty1 (Alt+P)"
      type="button"
    >
      /dev/tty1
    </button>
  </div>
  <div
    className="core-serial-port-indicator off"
  />
</div>
`;
